website/www/site/content/en/documentation/patterns/state-and-timers.md

```diff
@@ -0,0 +1,143 @@
+- a TimerSpec which will callback at a certain point in time - in this case at the end of the window.
+- the `process`-function will add newly incoming events to the state and set the timer to the end of the window.
+- the `expiry`-function in Python `onTimer`-function in Java will read the state on expiry and output batches of the data ready to do a remote external service in the next transform.
+
+The result in Python:
+
+```python
+class _BatchItems(DoFn):
+    elements_bag = BagStateSpec('buffer', IterableCoder(PickleCoder()))
+    event_timer = TimerSpec('event_timer', TimeDomain.WATERMARK)
+
+    def __init__(self, max_batch_size=5000):
+        self.max_batch_size = max_batch_size
+
+    def process(self,
+                element,
+                window=DoFn.WindowParam,
+                buffer_state=DoFn.StateParam(elements_bag),
+                event_timer=DoFn.TimerParam(event_timer)):
+
+        buffer_state.add(element)
+        event_timer.set(window.max_timestamp())
+
+    @on_timer(event_timer)
+    def expiry(self, buffer_state=DoFn.StateParam(elements_bag), event_timer=DoFn.TimerParam(event_timer)):
+        batch = [element for element in buffer_state.read()]
+        if not batch:
+            return
+
+        key, _ = batch[0]
+        current_batch_size = itertools.count()
+        output = []
+
+        if batch:
+            for event in batch:
+                clear_buffer = next(current_batch_size) >= self.max_batch_size
+                if clear_buffer:
+                    yield (key, output)
+                    output.clear()
+                    current_batch_size = itertools.count()
+                output.append(event[1])
+
+        if len(output) > 0:
+            yield (key, output)
+```
+
+and in Java:
+
+```java
+class BatchRequestForRecommendationAI extends DoFn<KV<String, GenericJson>, KV<String, Iterable<GenericJson>>> {
+    private static final Logger LOG = LoggerFactory.getLogger(BatchRequestForRecommendationAI.class);
+
+    private final Counter numberOfRowsBagged =
+            Metrics.counter(BatchRequestForRecommendationAI.class, "numberOfRowsBagged");
+
+    private final Integer maxBatchSize;
+
+    @StateId("elementsBag")
+    private final StateSpec<BagState<KV<String, GenericJson>>> elementsBag = StateSpecs.bag();
+
+    @TimerId("eventTimer")
+    private final TimerSpec eventTimer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    public BatchRequestForRecommendationAI(Integer maxBatchSize) {
+        this.maxBatchSize = maxBatchSize;
+    }
+
+    @ProcessElement
+    public void process(
+            @Element KV<String, GenericJson> element,
+            @StateId("elementsBag") BagState<KV<String, GenericJson>> elementsBag,
+            @TimerId("eventTimer") Timer eventTimer,
+            BoundedWindow w) {
+        elementsBag.add(element);
+        eventTimer.set(w.maxTimestamp());
```

Comment:
What if we are in Global window with process time trigger per element. 
Also if this is a long window, would we not want output before the end of the window? Maybe this could be something that is configured by the user, so that ever x secs the buffer is checked?

Which would also make the code below also need to optimise so that we dont need to read the whole bag ever time.